### PR TITLE
[WIP] Prototype of DataJointWorker and WorkerLog

### PR DESCRIPTION
**Pull request by ttngu207**

### CODE REVIEW

File: workflow/djworker.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    process_timestamp : datetime      # timestamp of the processing job
+    process           : varchar(64)
+    ---
+    worker_name=''    : varchar(255)  # name of the worker
+    host              : varchar(255)  # system hostname
+    user=''           : varchar(255)  # database user
+    pid=0             : int unsigned  # system process id  
+    """
+
+    _table_name = '~WorkerLog'
```
Comment (dimitri-yatsenko) [inline workflow/djworker.py:27]: we use lower snake case for all table names
```suggestion
    _table_name = '~worker_log'
```

File: workflow/djworker.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        if isinstance(process, dj.user_tables.TableMeta):
+            schema_name, table_name = process.full_table_name.split('.')
+            schema_name = schema_name.strip('`').replace(db_prefix, '')
+            table_name = dj.utils.to_camel_case(table_name.strip('`'))
+            process_name = f'{schema_name}.{table_name}'
+            user = process.connection.get_user()
+        elif inspect.isfunction(process):
+            process_name = process.__name__
+            user = ''
+
```
Comment (dimitri-yatsenko) [inline workflow/djworker.py:40]: else raise error?

File: workflow/djworker.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        cls.insert1({'process': process_name,
+                     'process_timestamp': datetime.utcnow(),
+                     'worker_name': worker_name,
+                     'host': platform.node(),
+                     'user': user,
+                     'pid': os.getpid()})
+
+    @classmethod
+    def print_recent_jobs(cls, backtrack_minutes=60):
```
Comment (dimitri-yatsenko) [inline workflow/djworker.py:55]: This function does not print.  So change the name.
```suggestion
    def recent_jobs(cls, backtrack_minutes=60):
```

File: workflow/djworker.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        return recent_jobs
+
+    @classmethod
+    def delete_old_logs(cls, cutoff_days=3):
+        old_jobs = (cls.proj(
+            elapsed_days=f'TIMESTAMPDIFF(DAY, process_timestamp, "{datetime.utcnow()}")')
+                    & f'elapsed_days > {cutoff_days}')
+        if old_jobs:
+            with dj.config(safemode=False):
+                (cls & old_jobs).delete()
```
Comment (dimitri-yatsenko) [inline workflow/djworker.py:75]: consider using `delete_quick`
```suggestion
                (cls & old_jobs).delete_quick()
```

File: workflow/djworker.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        elif inspect.isfunction(process):
+            self._processes_to_run.append(('function', process, kwargs))
+        else:
+            raise NotImplemented(f'Unable to handle processing step of type {type(process)}')
+
+    def run(self):
+        start_time = time.time()
+        while ((time.time() - start_time < self._run_duration)
+               or (self._run_duration is None)
+               or (self._run_duration < 0)):
```
Comment (dimitri-yatsenko) [inline workflow/djworker.py:113]:
```suggestion
        while (time.time() - start_time < self._run_duration
               or self._run_duration is None
               or self._run_duration < 0):
```

File: workflow/djworker.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+            for process_type, process, kwargs in self._processes_to_run:
+                WorkerLog.log_process_job(process, worker_name=self.name,
+                                          db_prefix=self._db_prefix)
+                if process_type == 'dj_table':
+                    process.populate(**{**_populate_settings, **kwargs})
```
Comment (dimitri-yatsenko) [inline workflow/djworker.py:119]:
```suggestion
                    process.populate(**_populate_settings, **kwargs)
```

File: workflow/process.py [hunk 7]
```diff
@@ -1,66 +1,48 @@
...
+            logger.debug(str(e))
+        except Exception as e:
+            logger.error(str(e))
 
 
-def _clean_up():
-    (ephys.schema.jobs & 'status = "error"'
-     & [f'error_message LIKE "{e}"'
-        for e in _generic_errors + ['%FileNotFound%']]).delete()
+def auto_generate_clustering_tasks():
```
Comment (dimitri-yatsenko) [inline workflow/process.py:20]: I suggest using simplifying to `generate_clustering_tasks`

File: workflow/process.py [hunk 8]
```diff
@@ -1,66 +1,48 @@
...
+from workflow.djworker import WorkerLog, DataJointWorker
 
 logger = logging.getLogger(__name__)
 logger.setLevel('INFO')
 
 
-_generic_errors = ["%Deadlock%", "%DuplicateError%", "%Lock wait timeout%",
-                   "%MaxRetryError%", "%KeyboardInterrupt%",
-                   "InternalError: (1205%", "%SIGTERM%"]
+def auto_generate_probe_insertions():
```
Comment (dimitri-yatsenko) [inline workflow/process.py:10]: Consider simplifying to `generate_probe_insertions`

**Review by dimitri-yatsenko — CHANGES_REQUESTED** [review]

**Comment by guzman-raphael** [comment]
@ttngu207 Now that we have [datajoint-utilities](https://github.com/vathes/datajoint-utilities), would you mind moving this there?

**Comment by ttngu207** [comment]
Closing this PR, moving this `DataJointWorker` mechanism to: https://github.com/vathes/datajoint-utilities/pull/4

**Pull request closed without merging**